packages/@aws-cdk/aws-elasticsearch/lib/domain.ts

```diff
@@ -748,90 +820,494 @@ export class Domain extends cdk.Resource implements IDomain {
+    const versionNumber = parseFloat(elasticsearchVersion.toString());
+    const encryptionAtRestEnabled = props.encryptionAtRestOptions?.enabled ?? (props.encryptionAtRestOptions?.kmsKey != null);
+    const ebsEnabled = props.ebsOptions != null;
+
+    const isInstanceType = function(instanceType: string) : Boolean {
+      return masterInstanceType.startsWith(instanceType) || dataInstanceType.startsWith(instanceType);
+    };
+
+    const isSomeInstanceType = function(...instanceTypes: string[]) : Boolean {
+      return instanceTypes.some(isInstanceType);
+    };
+
+    const isEveryInstanceType = function(...instanceTypes: string[]) : Boolean {
+      return instanceTypes.every(isInstanceType);
+    };
+
+    // Validate feature support for the given Elasticsearch version, per
+    // https://docs.aws.amazon.com/elasticsearch-service/latest/developerguide/aes-features-by-version.html
+    if (versionNumber < 5.1) {
+      if (
+        props.logPublishingOptions?.slowIndexLogEnabled
+        || props.logPublishingOptions?.appLogEnabled
+        || props.logPublishingOptions?.slowSearchLogEnabled
+      ) {
+        throw new Error('Error and slow logs publishing requires Elasticsearch version 5.1 or later.');
+      }
+      if (props.encryptionAtRestOptions?.enabled) {
+        throw new Error('Encryption of data at rest requires Elasticsearch version 5.1 or later.');
+      }
+      if (props.cognitoOptions != null) {
+        throw new Error('Cognito authentication for Kibana requires Elasticsearch version 5.1 or later.');
+      }
+      if (isSomeInstanceType('c5', 'i3', 'm5', 'r5')) {
+        throw new Error('C5, I3, M5, and R5 instance types require Elasticsearch version 5.1 or later.');
+      }
+    } else if (versionNumber < 5.3) {
+      if (props.automatedSnapshotStartHour) {
+        throw new Error('Hourly automated snapshots requires Elasticsearch version 5.3 or later.');
+      }
+    } else if (versionNumber < 6.0) {
+      if (props.nodeToNodeEncryptionEnabled) {
+        throw new Error('Node-to-node encryption requires Elasticsearch version 6.0 or later.');
+      }
+    }
+
+    // Validate against instance type restrictions, per
+    // https://docs.aws.amazon.com/elasticsearch-service/latest/developerguide/aes-supported-instance-types.html
+    if (isInstanceType('i3') && ebsEnabled) {
+      throw new Error('I3 instance types do not support EBS storage volumes.');
+    }
+
+    if (isSomeInstanceType('m3', 'r3', 't2') && encryptionAtRestEnabled) {
+      throw new Error('M3, R3, and T2 instance types do not support encryption of data at rest.');
+    }
+
+    if (isInstanceType('t2.micro') && versionNumber > 2.3) {
+      throw new Error('The t2.micro.elasticsearch instance type supports only Elasticsearch 1.5 and 2.3.');
+    }
+
+    // Only R3 and I3 support instance storage, per
+    // https://aws.amazon.com/elasticsearch-service/pricing/
+    if (!ebsEnabled && !isEveryInstanceType('r3', 'i3')) {
```

Comment:
I don't see how this works.

`isEveryInstanceType('r3', 'i3')` I don't think this can ever be true since it would require `isIntanceType('r3')` && `isInstanceType('i3')` which is not possible.

Reply:
Yeah, my `isEveryInstanceType` was wrong!